src/ray/core_worker/profiling.cc

```diff
@@ -0,0 +1,58 @@
+#include <chrono>
+
+#include "ray/core_worker/profiling.h"
+
+namespace ray {
+
+namespace worker {
+
+ProfilingEvent::ProfilingEvent(Profiler &profiler, const std::string &event_type)
+    : profiler_(profiler) {
+  inner_.set_event_type(event_type);
+  inner_.set_start_time(current_sys_time_seconds());
+}
+
+Profiler::Profiler(WorkerContext &worker_context, const std::string &node_ip_address,
+                   std::unique_ptr<gcs::RedisGcsClient> &gcs_client)
+    : worker_context_(worker_context), gcs_client_(gcs_client) {
+  profile_info_.set_component_type(WorkerTypeString(worker_context_.GetWorkerType()));
+  profile_info_.set_component_id(worker_context_.GetWorkerID().Binary());
+  profile_info_.set_node_ip_address(node_ip_address);
+  thread_ = std::thread(&Profiler::PeriodicallyFlushEvents, this);
```

Comment:
@edoakes and I just discovered that this is buggy, that `PeriodicallyFlushEvents` tries to grab a mutex, which hasn't been initialized yet.